src/t_zset.c

```diff
@@ -119,16 +119,35 @@ void zslFree(zskiplist *zsl) {
     zfree(zsl);
 }
 
+#define RANDOM_CTZ_MAX 31
+#define RANDOM_MAX 2147483647
+
 /* Returns a random level for the new skiplist node we are going to create.
  * The return value of this function is between 1 and ZSKIPLIST_MAXLEVEL
  * (both inclusive), with a powerlaw-alike distribution where higher
  * levels are less likely to be returned. */
 int zslRandomLevel(void) {
-    static const int threshold = ZSKIPLIST_P*RAND_MAX;
+#if ZSKIPLIST_P_DENOMINATOR == 4 && ZSKIPLIST_MAXLEVEL == 32
+    /* Optimized implementation for P = 1/4 and MAXLEVEL = 32.
+     * In most cases, only a single random number is generated, enhancing performance.
+     * Calculate the count of trailing zeros in a random number. The random number is
+     * bitwise OR'ed with 1<<MAX_RANDOM_CTZ to ensure it's not zero. */
+    int ctz = __builtin_ctz(random()|(1<<RANDOM_CTZ_MAX));
+    /* If ctz is 31 (all bits are zero), repeat the process to increase level.
+     * This is an extremely rare case. */
+    if (ctz == RANDOM_CTZ_MAX)
+       ctz += __builtin_ctz(random()|(1<<RANDOM_CTZ_MAX));
```

Comment:
```suggestion
    int ctz = __builtin_ctzll(genrand64_int63()|(1<<63));
```
Is it same result?

Reply:
Yes! Thanks for the suggestion.
 I've looked into it and found that it's indeed an in-built implementation within Redis, which will make it more elegant.